src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,155 @@
-        }
-    
-        /*=================================GetCodePageFromName==========================
-        **Action: Given a encoding name, return the correct code page number for this encoding.
-        **Returns: The code page for the encoding.
-        **Arguments:
-        **  name    the name of the encoding
-        **Exceptions:
-        **  ArgumentNullException if name is null.
-        **  internalGetCodePageFromName will throw ArgumentException if name is not a valid encoding name.
-        ============================================================================*/
-        
-        internal static int GetCodePageFromName(String name)
+        internal static EncodingInfo[] GetEncodings()
         {
-            // TODO: Implement this fully.
-            switch (name)
-            {
-                case "utf-7":
-                    return CodePageUtf7; 
+            EncodingInfo[] arrayEncodingInfo = new EncodingInfo[s_encodingDataTableItems.Length];
 
-                case "utf-8":
-                    return CodePageUtf8;
+            for (int i = 0; i < s_encodingDataTableItems.Length; i++)
+            {
+                CodePageDataItem dataItem = s_encodingDataTableItems[i];
 
-                case "utf-16":
-                    return CodePageUtf16;
+                arrayEncodingInfo[i] = new EncodingInfo(dataItem.CodePage, dataItem.WebName,
+                    Environment.GetResourceString("Globalization.cp_" + dataItem.CodePage));
+            }
 
-                case "utf-16BE":
-                    return CodePageUtf16BE;
+            return arrayEncodingInfo;
+        }
 
-                case "utf-32":
-                    return CodePageUtf32;
+        internal static int GetCodePageFromName(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException("name");
+            }
+            Contract.EndContractBlock();
 
-                default:
-                    return CodePageUtf8;
+            ushort codePage;
+            if (s_encodingDataTable.TryGetValue(name, out codePage))
+            {
+                return codePage;
             }
+
+            // The encoding name is not valid.
+            throw new ArgumentException(
+                string.Format(
+                    CultureInfo.CurrentCulture,
+                    Environment.GetResourceString("Argument_EncodingNotSupported"), name), "name");
         }
-    
-        [System.Security.SecuritySafeCritical]  // auto-generated
-        unsafe internal static CodePageDataItem GetCodePageDataItem(int codepage) {
-            // TODO: Implement this fully.
+
+        internal static CodePageDataItem GetCodePageDataItem(int codepage)
+        {
             switch (codepage)
             {
-                case CodePageUtf7:
-                    return new CodePageDataItem("utf-7", CodePageUtf7, "utf-7", "utf-7", 771);
+                case 1200: return s_encodingDataTableItems[0];
```

Comment:
I know I suggested the switch as it is here, but it'd be nice if there was a Contract.Assert to verify the CodePage matches the item from the array. Sorry for not including that in my original feedback.

Maybe something like...

(Typing this from my phone...)

```c#
CodePageDataItem item = null;
switch (codepage)
{
    case 1200:
        item = s_encodingDataTableItems[0];
        break;

    ...etc...
}

Contract.Assert(item != null ? item.CodePage == codepage : true);

return item;
```